### ISSUE TITLE
Array access in interrupt handler seems to corrupt concurrent array accesses

### ISSUE THREAD

**Issue by ToboterXP** [cc65/cc65#2217]
I'm on CX16, and currently writing a game. I noticed a lot of very weird issues, where some variables seemed to temporarily corrupt for no reason while the keyboard was used. This is the code snippet (keyDown is just a 128 byte array).
![image](https://github.com/cc65/cc65/assets/55210408/7544abad-e331-45ff-bb5e-566baf154dd7)
After I rewrote this code in inline assembly the issues stopped.

Apparently some of the internal functions cc65 uses in its compilation aren't concurrency safe, and break if used in an interrupt handler.

**Comment by polluks** [comment]
Please show the compiler's output with -S and --add-source

**Comment by ToboterXP** [comment]
![grafik](https://github.com/cc65/cc65/assets/55210408/48adf05c-b998-491f-accd-3b32cdd05d8e)
Here you go.

**Comment by ToboterXP** [comment]
Though I did it with an & 0b10000000 once, rather than a shift, and the problem still persisted. So the asr functions are probably not at fault.

**Comment by mrdudz** [comment]
You'll have to provide a complete example.... Are you using set_irq to set up the interrupt handler?

**Comment by ToboterXP** [comment]
![grafik](https://github.com/cc65/cc65/assets/55210408/7ecabc31-0b99-489e-b4c6-c1b49888e9fb)
No, this.

**Comment by mrdudz** [comment]
There is your problem :)

**Comment by ToboterXP** [comment]
How would I use set_irq for something like this? This isn't just the normal irq.

**Comment by ToboterXP** [comment]
It's specifically the keyhandler callback used by the X16

**Comment by ToboterXP** [comment]
Though I think it works very similarly on all Commodore machines

**Comment by mrdudz** [comment]
The short answer is... you can't. The only provided mechanism to handle interrupts in C is via set_irq.

The long answer is: you'll have to write a custom wrapper in assembly, which saves all zeropage addresses used by the runtime, then calls your C function, and restores those zeropage locations.

**Comment by ToboterXP** [comment]
Then I guess that justifies making this a bug report.

**Comment by mrdudz** [comment]
There is no bug - it works as intended.

**Comment by ToboterXP** [comment]
It is a bug. There's no indication it shouldn't work anywhere, and it's a core feature of the X16. If you don't support it you specifically have to document it.

Am 7. Oktober 2023 21:39:59 MESZ schrieb Bob Andrews ***@***.***>:
>There is no bug - it works as intended.
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/cc65/cc65/issues/2217#issuecomment-1751802656
>You are receiving this because you authored the thread.
>
>Message ID: ***@***.***>
-- 
Diese Nachricht wurde von meinem Android-Gerät mit K-9 Mail gesendet.

**Comment by mrdudz** [comment]
> It is a bug. There's no indication it shouldn't work anywhere, and it's a core feature of the X16. If you don't support it you specifically have to document it.

Its not a bug. It has nothing to with the X16 either. Its just how every compiler and every runtime works. Generally C does not even support interrupts, its always a compiler and target specific thing. You can't expect a compiled function to work as is when its called by random system vectors. You always have to save and restore whatever the environment uses on a given target for that to work, and often also tell the compiler to produce a different kind of function that is specifically suited as an interrupt handler (cc65 does not have such feature though).

Actually that code has more problems - the compiler may emit a function prolog and epilog at the start and end of the function, which means your inline asm may or may not work, depending on what else is used in that function (it only works right now because that function does almost nothing).

However, "how to use interrupts in C" is surely something that needs more documentation (and perhaps an example or two).

**Comment by ToboterXP** [comment]
I mean, C is fundamentally intended to be concurrency-safe. It was literally designed to make Unix, a multithreading based OS. And the ability to deal with being interrupted (with signals, callback handlers or preemptive multithreading) is fundamental to C programming (or programming in general). Generally not supporting it is highly unusual and should be clearly marked. 

Am 8. Oktober 2023 01:12:03 MESZ schrieb Bob Andrews ***@***.***>:
>> It is a bug. There's no indication it shouldn't work anywhere, and it's a core feature of the X16. If you don't support it you specifically have to document it.
>
>Its not a bug. It has nothing to with the X16 either. Its just how every compiler and every runtime works. Generally C does not even support interrupts, its always a compiler and target specific thing. You can't expect a compiled function to work as is when its called by random system vectors. You always have to save and restore whatever the environment uses on a given target for that to work, and often also tell the compiler to produce a different kind of function that is specifically suited as an interrupt handler (cc65 does not have such feature though).
>
>Actually that code has more problems - the compiler may emit a function prolog and epilog at the start and end of the function, which means your inline asm may or may not work, depending on what else is used in that function (it only works right now because that function does almost nothing).
>
>However, "how to use interrupts in C" is surely something that needs more documentation (and perhaps an example or two).
>
>-- 
>Reply to this email directly or view it on GitHub:
>https://github.com/cc65/cc65/issues/2217#issuecomment-1751850615
>You are receiving this because you authored the thread.
>
>Message ID: ***@***.***>
-- 
Diese Nachricht wurde von meinem Android-Gerät mit K-9 Mail gesendet.

**Comment by kugelfuhr** [comment]
> I mean, C is fundamentally intended to be concurrency-safe.

No. Exactly the opposite is true. The C standard doesn't make *any* guarantees in this respect. It does even say in ISO/IEC 9899:1999 (E) 7.1.4p4: 

"The functions in the standard library are not guaranteed to be reentrant and may modify objects with static storage duration."

> It was literally designed to make Unix, a multithreading based OS.

This is also false. When C was designed, Unix did exist but wasn't multithreaded. The Unix from that time was capable of multitasking, but that was implemented by different processes, not threads. Code within one process didn't need to be reentrant (with the exception of signals).

> And the ability to deal with being interrupted (with signals, callback handlers or preemptive multithreading) is fundamental to C programming (or programming in general).

You might try looking up "interrupts", "callback handlers" or "preemptive multithreading" in the C standard. All these things aren't even mentioned and therefore out of the scope of standard C. Compilers may implement features to use these things, like cc65 does with `set_irq()`, but this is not required, always implementation specific, and not guaranteed to be available.

Signals are part of the C standard and they can be used safely with cc65 (with the limitations given in the standard).

**Comment by mrdudz** [comment]
> Generally not supporting it is highly unusual and should be clearly marked.

It is actually not uncommon, on embedded targets without an OS, that you have to provide certain boilerplate code and/or specifically mark functions as interrupt handlers. Because often such a handler must preserve eg more registers than a regular function - which is pretty much what is needed with cc65.

see eg https://developer.arm.com/documentation/dui0774/l/Compiler-specific-Function--Variable--and-Type-Attributes/--attribute----interrupt--type-----function-attribute

or here: https://www.analog.com/en/app-notes/using-the-free-sdcc-c-compiler-to-develop-firmware-for-the-ds89c430450-family-of-microcontrollers.html

(just two random hits i found with a quick google search - many other targets have similar requirements)

Something like "attribute interrupt", or perhaps a calling convention that serves the same purpose, is certainly something that is missing in cc65 when comparing it with other compilers that target small embedded systems - but as said, it doesnt exist yet.

**Comment by mrdudz** [comment]
Moving this to discussions
